bindings/azure/signalr/signalr.go

```diff
@@ -226,41 +245,114 @@ func (s *SignalR) sendMessageToSignalR(ctx context.Context, url string, token st
 
 	resp, err := s.httpClient.Do(httpReq)
 	if err != nil {
-		return fmt.Errorf("request to azure signalr api failed: %w", err)
+		return nil, fmt.Errorf("request to azure signalr api failed: %w", err)
 	}
 	defer resp.Body.Close()
 
 	// Read the body regardless to drain it and ensure the connection can be reused
 	body, err := io.ReadAll(resp.Body)
 	if err != nil {
-		return err
+		return nil, err
 	}
 
 	if resp.StatusCode != http.StatusOK && resp.StatusCode != http.StatusAccepted {
-		return fmt.Errorf("azure signalr failed with code %d, content is '%s'", resp.StatusCode, string(body))
+		return nil, fmt.Errorf("azure signalr failed with code %d, content is '%s'", resp.StatusCode, string(body))
 	}
 
 	s.logger.Debugf("azure signalr call to '%s' completed with code %d", url, resp.StatusCode)
 
-	return nil
+	return body, nil
 }
 
 func (s *SignalR) Operations() []bindings.OperationKind {
-	return []bindings.OperationKind{bindings.CreateOperation}
+	return []bindings.OperationKind{bindings.CreateOperation, ClientNegotiateOperation}
 }
 
 func (s *SignalR) Invoke(ctx context.Context, req *bindings.InvokeRequest) (*bindings.InvokeResponse, error) {
+	if req.Operation == ClientNegotiateOperation {
+		return s.GenerateClientNegotiateResponse(ctx, req)
+	} else {
+		return s.SendMessages(ctx, req)
+	}
+}
+
+func (s *SignalR) GenerateClientNegotiateResponse(ctx context.Context, req *bindings.InvokeRequest) (*bindings.InvokeResponse, error) {
+	// Generate token
+	hub, err := s.getHub(req)
+
+	if err != nil {
+		return nil, err
+	}
+
+	userId, ok := req.Metadata[userKey]
+	if !ok {
+		userId = ""
+	}
+
+	clientUrl := fmt.Sprintf("%s/client/?hub=%s", s.endpoint, hub)
+
+	// If we have an Azure AD token provider, invoke REST API to generate token
+	// Otherwise, generate token locally
+	var token string
+	if s.aadToken != nil {
+		token, err = s.GetAadClientAccessToken(ctx, hub, userId)
+	} else {
+		// Default to 60 minutes
+		token, err = s.getToken(ctx, clientUrl, userId, 60)
+	}
+
+	if err != nil {
+		return nil, fmt.Errorf("Error generating negotaite payload")
+	}
+
+	// Create the negotiate JSON payload
+	payload := map[string]string{
+		"url":         clientUrl,
+		"accessToken": token,
+	}
+	data, err := json.Marshal(payload)
+	if err != nil {
+		return nil, fmt.Errorf("Error generating negotaite payload")
+	}
+	contentType := "application/json"
+	response := bindings.InvokeResponse{
+		Data:        data,
+		Metadata:    map[string]string{},
+		ContentType: &contentType,
+	}
+	return &response, nil
+}
+
+func (s *SignalR) GetAadClientAccessToken(ctx context.Context, hub string, userId string) (string, error) {
+	aadToken, err := s.getAadToken(ctx)
+	if err != nil {
+		return "", err
+	}
+
+	url := fmt.Sprintf("%s/api/hubs/%s/:generateToken?userId=%s&api-version=%s", s.endpoint, hub, userId, apiVersion)
```

Comment:
Is there any injection risk here? Like customer uses `userId=&api-version=badass#`?